### DISCUSSION TITLE
Implementing own ASreview plugin for own dataset extensions

### DISCUSSION THREAD

**Discussion by TheB4ssPl4y3r** [asreview/asreview#1001] [Q&A] [answered]
Hei everyone

For a project I try to implement a dataset extension plugin to ASReview. I took the official template (https://github.com/asreview/template-extension-new-dataset) and tried to adjust it for our dataset (https://github.com/TheB4ssPl4y3r/lights_test.git). I managed to see the plugin in the ASReview plug-in interface (instalation of the package was successfully), but the dataset does not load and an error occured in my terminal. I tried then to just install the package as it is in the template to make sure I did no wrong changes in the template. I uninstalled the non-functioning package and installed the raw template package. Worked fine and the plugin pops up in the plug-in interface of ASReview. But I got the same problem here as mentioned before. Data starts loading but does not finish loading and in my terminal this error pops up:

line 375, in _parse
    raise ValueError("unknown url type: %r" % self.full_url)
ValueError: unknown url type: ''
ERROR:root:unknown url type: ''

<img width="773" alt="image" src="https://user-images.githubusercontent.com/83454695/156739987-bbe25b94-82a2-49a3-99f2-a829b66e0a7c.png">

I also installed the covid-19 extension and tried to load the date from the covid-19 plug in. The loading took some time (big dataset) but it does work without errors.

Does anyone has experience with building simple dataset extension plug-ins and can give me a hit what I am doing wrong here?
I also think the documentation does not help to much here as it is not very detailed and specific, so I thought the possibility of using a template is a very nice opportunity to easy implement an own plug-in. Not for me I think ;)

**Comment by jteijema** [comment]
Hi @TheB4ssPl4y3r,

I'm currently working on fixing the dataset plugin! It should be available next week, and then it will be usable for the latest version of ASReview.

Jelle

**Comment by Rensvandeschoot** [comment]
@TheB4ssPl4y3r Great you are working with the template to create a plug-in with your own data!!

I have successfully used the template in version 0.17 to process [demo-data sets containing emails](https://github.com/asreview-ftm-hackathon/ASReview-plugin-FTM) for our [hackathon](https://asreview.nl/blog/asreview-hackathon-for-follow-the-money/). Can you try downgrading to version 0.18 to see if your plug-in works?

``` 
pip install asreview==0.17
```

**Reply by TheB4ssPl4y3r** [reply]
@Rensvandeschoot

Thanks for your input. Unfortunately I get another Error in version 0.18 (ERROR:root:Dataset undefined not found) as with version 0.19.1 (ERROR:root:unknown url type: ''). The error now occurs for both, the template and the covid plugins.


- asreview 0.19.1 the covid plugins work. The untouched extension template, installed via 'pip install git+https://github.com/asreview/template-extension-new-dataset.git' successfully (plugin in the plugin interface), does lead to the error I mentioned before: 
request.py", line 375, in _parse
    raise ValueError("unknown url type: %r" % self.full_url)
ValueError: unknown url type: ''
ERROR:root:unknown url type: ''

- asreview 0.18  the covid plugins do not work and lead to this error:
datasets.py", line 301, in find
    raise DataSetNotFoundError(
asreview.datasets.DataSetNotFoundError: Dataset undefined not found
ERROR:root:Dataset undefined not found

- asreview 0.18 the untouched template, installed via 'pip install git+https://github.com/asreview/template-extension-new-dataset.git' successfully (plugin in the plugin interface), does lead to an error:
datasets.py", line 301, in find
    raise DataSetNotFoundError(
asreview.datasets.DataSetNotFoundError: Dataset undefined not found
ERROR:root:Dataset undefined not found

best,
Ramon

**Reply by Rensvandeschoot** [reply]
This error relates to https://github.com/asreview/asreview/issues/830 and was solved in https://github.com/asreview/asreview/pull/866. So, we are back to your original issue....

**Reply by TheB4ssPl4y3r** [reply]
I tried also version 0.17 as you updated in your first answer and got just a white screen when clicking on the plug in tab.   No error occured in my terminal.

So I think I just wait until the template is updated

Thx best,
Ramon

**Comment by jteijema** [comment]
Hi @TheB4ssPl4y3r, I'm looking into this issue again. And I think the template should in fact work for version 0.19, the latest version. Could you share your implementation again? I'd like to replicate the issue.

**Reply by jteijema** [reply]
`ValueError: unknown url type: '' ` is the error that results from an empty URL string:
 
![image](https://user-images.githubusercontent.com/28191416/159900936-8fec2eb4-23a9-4d60-ab7f-3f03aab9710e.png)

**Reply by TheB4ssPl4y3r** [reply]
Hi @JTeijema 

I start over and try to explain my hole implementation and errors that eventually occur.

- First I uninstalled the template extension package. So I start with these asreview packages:

<img width="882" alt="image" src="https://user-images.githubusercontent.com/83454695/160129308-41815b21-3a8d-426f-b879-88bf2197a206.png">
ASReview does work as it should with these packages installed.

- I used the official template and added it to my repos in public mode (https://github.com/TheB4ssPl4y3r/template-extension-new-dataset.git).
- I changed the your_dataset.py file (https://github.com/TheB4ssPl4y3r/template-extension-new-dataset/blob/main/asreviewcontrib/dataset_name/your_dataset.py) and added the url of the default dataset your_dataset.csv (https://github.com/TheB4ssPl4y3r/template-extension-new-dataset/blob/main/data/your_dataset.csv)
<img width="805" alt="image" src="https://user-images.githubusercontent.com/83454695/160133644-8982a903-5ae7-4d2a-9de4-2424fe08baba.png">

- I installed the adjusted package via 'pip install git+https://github.com/TheB4ssPl4y3r/template-extension-new-dataset.git'
<img width="855" alt="image" src="https://user-images.githubusercontent.com/83454695/160135398-e42b2daf-af62-4d94-a8ff-fb3a36fadf63.png">

- I started ASreview and tried to open the data from the plugins tab where the 'Your Data' is displayed as it should.
<img width="805" alt="image" src="https://user-images.githubusercontent.com/83454695/160136392-44acee27-35e5-44fd-bfbd-b66758745519.png">

- An error occurred:
<img width="824" alt="image" src="https://user-images.githubusercontent.com/83454695/160137201-e9ba5257-640e-42d1-a8e6-6831033fafe1.png">

The url error disappeared, great but I doo have another one now.

Thanks and nice weekend
best,

TheB4ssPl4y3r

**Reply by J535D165** [reply]
Please change the URL of your dataset to https://raw.githubusercontent.com/TheB4ssPl4y3r/template-extension-new-dataset/main/data/your_dataset.csv (click on `Raw` in your link ;)) 

Please keep in mind that this dataset only contains 1 record. ASReview might struggle with it.

**Reply by TheB4ssPl4y3r** [reply]
great it worked thank you very much @J535D165 . Im gonna try it the next week with our own dataset.

Have a nice one

**Comment by Rensvandeschoot** [comment, marked as answer]
See the [reply ](https://github.com/asreview/asreview/discussions/1001#discussioncomment-2439343)by @J535D165 for the answer
